### PR TITLE
INSTALLER: 16.04 and Kinetic integration.

### PR DESCRIPTION
**Pull request by RustyBamboo**
- Installs Catkin, Rosdep, binutils, Kinetic-desktop-full, OMPL, spacenav-node
- Downgrades Google's protobuf to compile Gazebo7's defaulted 2.6 protoc messages

Creating a pull request to encourage participation in developing and testing installer.

### CODE REVIEW

File: install.sh [hunk 1]
```diff
@@ -1,81 +1,155 @@
...
+    NET_CHECK=false
+    echo -n "[ " && instfail && echo -n "] "
+fi
+echo "Internet connectivity check"
+
+if !($OS_CHECK); then
+
+    # The script will not allow the user to install on an unsupported OS
     instwarn "Terminating installation due to incorrect OS (detected $DTETCTED_OS)"
+    instwarn "MIL projects require Ubuntu 14.04 (trusty)"
```
Comment (DSsoto) [inline install.sh]: This line is no longer applicable.

File: install.sh [hunk 2]
```diff
@@ -1,81 +1,155 @@
...
 sudo apt-get update -qq
-if (env | grep --quiet "SUB=true"); then
-    sudo apt-get install -qq ros-indigo-desktop-full
-else
-    sudo apt-get install -qq ros-indigo-desktop
-fi
-sudo apt-get install -qq python-catkin-pkg python-rosdep gazebo7
+sudo apt-get install -qq catkin python-rosdep
+
+instlog "Installing ROS Kinetic base packages"
```
Comment (DSsoto) [inline install.sh]: This is a desktop full install, not base packages. Let's not mislead people.

File: install.sh [hunk 3]
```diff
@@ -85,55 +159,92 @@ if !([ -f /etc/ros/rosdep/sources.list.d/20-default.list ]); then
...
-
-#=====================================#
-# Workspace and Sub Repository Set Up #
-#=====================================#
+instalog "Installing Sub8 ROS dependencies"
+sudo apt-get install -qq binutils-dev
+sudo apt-get install -qq ros-kinetic-ompl
+sudo apt-get install -qq ros-kinetic-spacenav-node
+
+# Get older version of Google protobuf for Gazebo
```
Comment (DSsoto) [inline install.sh:178]: Have you confirmed with others that this is necessary, or the best way to do it?

File: install.sh [hunk 4]
```diff
@@ -85,55 +159,92 @@ if !([ -f /etc/ros/rosdep/sources.list.d/20-default.list ]); then
...
 else
     instlog "Using existing catkin workspace at $CATKIN_DIR"
+    cd $CATKIN_DIR/src
 fi
 
-# If we're in the Semaphore-ci, we should run catkin_make in the actual build thread
-if (env | grep --quiet "SEMAPHORE=true"); then
-    mv ~/Sub8 "$CATKIN_DIR/src"
-fi
+CMAKE_PREFIX_PATH=/opt/ros/kinetic
```
Comment (DSsoto) [inline install.sh]: Could this potentially overwrite something important? Wouldn't it be better to just append?

File: install.sh [hunk 5]
```diff
@@ -1,81 +1,155 @@
...
+
+            # Will return true if ping was successful and packet loss was below 25%
+            return `true`
+        fi
+    fi
+    return `false`
+}
+
+REQUIRED_OS="xenial"
+CATKIN_DIR=~/mil_ws
```
Comment (DSsoto) [inline install.sh:52]: Should we prompt the user and have this as the default?

**Review by DSsoto — COMMENTED** [review]
This is a good start. In my previous commit, I added a .sub_aliases file at the root of the repo, you should source this from the user's bashrc. I think we should also follow this configuration style for the other MIL repos.

We still need to find out how we're going to tie in CUDA into this. On machines not running already runing nvidia drivers this could be non-trivial.

File: install.sh (hunk 3 above)
Comment (RustyBamboo) [reply install.sh:178]: The gazebo messages are defaulted to protoc v2.6... The Ubuntu package is v3.0 (https://launchpad.net/ubuntu/+source/protobuf.) The different versions caused mine not to build. (Not sure if this is the case for others)

File: install.sh [hunk 6]
```diff
@@ -62,7 +62,7 @@ done
 
 # The paths to the aliases configuration files
 BASHRC_FILE=~/.bashrc
-ALIASES_FILE=$CATKIN_DIR/src/.sub_aliases
+ALIASES_FILE=$CATKIN_DIR/src/sub8/.sub_aliases
```
Comment (DSsoto) [inline install.sh]: The `s` in Sub8 should be capital.

File: install.sh [hunk 7]
```diff
@@ -1,81 +1,166 @@
...
-CATKIN_DIR=~/sub_ws
-REQUIRED_OS="trusty"
+LOGPREFIX="${LOGCOLOR}INSTALLER:"
+WARNPREFIX="${WARNCOLOR}ERROR:"
+PASSTEXT="${PASSCOLOR}PASS"
+FAILTEXT="${WARNCOLOR}FAIL"
 
 instlog() {
-    printf "$GOODPREFIX $@ $NOCOLOR\n"
+    printf "$LOGPREFIX $@ $NOCOLOR\n"
```
Comment (sentree) [inline install.sh:14]: I usually edit my bash scripts in nano, so naturally I prefer tabs to spaces. If there is a strong push to switch this to spaces, I recommend we use the bash standard of two spaces for indentation; however, I'd prefer to stick to tabs.

File: install.sh [hunk 8]
```diff
@@ -1,81 +1,166 @@
...
+        # Uses packet loss percentage to determine if the connection is strong
+        if [ $HOST_PING -lt 25 ]; then
+
+            # Will return true if ping was successful and packet loss was below 25%
+            return `true`
+        fi
+    fi
+    return `false`
+}
+
```
Comment (sentree) [inline install.sh:48]: Is there any particular reason you decided to do away with ros_git_get? Jake wrote the function to simplify the lines where git repositories were cloned and I was planning on using it's '--recursive' option to make our dependencies git submodules.

File: install.sh [hunk 9]
```diff
@@ -1,81 +1,166 @@
...
+# Confirmation of where to put CATKIN directory
+CATKIN_DIR=~/mil_ws
+while true; do
+    instlog "The CATKIN Directory is $CATKIN_DIR. Is this okay(y/n)? "
+    read yn
+    case $yn in
+        [Yy]* ) break;;
+        [Nn]* ) read -p "Enter a new directory: " CATKIN_DIR; break;;
+        * ) echo "Please answer yes or no.";;
```
Comment (sentree) [inline install.sh:59]: This could be confusing. Your code responds to values in [Y, y, N, n], but you request a 'yes' or 'no'.

File: install.sh [hunk 10]
```diff
@@ -1,81 +1,166 @@
...
+    fi
+    return `false`
+}
+
+REQUIRED_OS="xenial"
+
+# Confirmation of where to put CATKIN directory
+CATKIN_DIR=~/mil_ws
+while true; do
+    instlog "The CATKIN Directory is $CATKIN_DIR. Is this okay(y/n)? "
```
Comment (sentree) [inline install.sh:54]: Using a prompt based approach instead of script flags will probably work well for people that are new to MIL, but I will need to add a prompt for each flag in the common install script.

File: install.sh [hunk 11]
```diff
@@ -1,81 +1,166 @@
...
+    case $yn in
+        [Yy]* ) break;;
+        [Nn]* ) read -p "Enter a new directory: " CATKIN_DIR; break;;
+        * ) echo "Please answer yes or no.";;
+    esac
+done
+
+# The paths to the aliases configuration files
+BASHRC_FILE=~/.bashrc
+ALIASES_FILE=$CATKIN_DIR/src/Sub8/.sub_aliases
```
Comment (sentree) [inline install.sh:65]: It was a good idea to source the bash aliases file in the repository, but I will probably move some of these to software-common. I would also like to keep the file visible and in the scripts directory.

File: install.sh [hunk 12]
```diff
@@ -1,81 +1,166 @@
...
-if (env | grep --quiet "SUB=true"); then
-    sudo apt-get install -qq ros-indigo-desktop-full
-else
-    sudo apt-get install -qq ros-indigo-desktop
-fi
-sudo apt-get install -qq python-catkin-pkg python-rosdep gazebo7
+sudo apt-get install -qq catkin python-rosdep
+
+instlog "Installing the full ROS Kinetic"
+sudo apt-get install -qq ros-kinetic-desktop-full
```
Comment (sentree) [inline install.sh:157]: Good thinking here, we can finally do away with all of that metapackage nonsense!

File: install.sh [hunk 13]
```diff
@@ -85,55 +170,104 @@ if !([ -f /etc/ros/rosdep/sources.list.d/20-default.list ]); then
 fi
 rosdep update
 
+instlog "Installing Sub8 ROS dependencies"
```
Comment (sentree) [inline install.sh:173]: These should be under 'Dependency Installation'

File: install.sh [hunk 14]
```diff
@@ -85,55 +170,104 @@ if !([ -f /etc/ros/rosdep/sources.list.d/20-default.list ]); then
...
     instlog "Using existing catkin workspace at $CATKIN_DIR"
+    cd $CATKIN_DIR/src
 fi
 
-# If we're in the Semaphore-ci, we should run catkin_make in the actual build thread
-if (env | grep --quiet "SEMAPHORE=true"); then
-    mv ~/Sub8 "$CATKIN_DIR/src"
-fi
+CMAKE_PREFIX_PATH=$CMAKE_PREFIX_PATH:/opt/ros/kinetic
+git clone https://github.com/ros-drivers/driver_common
```
Comment (sentree) [inline install.sh:207]: What is this repository needed for? Also, I believe CMAKE_PREFIX_PATH is set when /opt/ros/kinetic/setup.bash is sourced.

File: install.sh [hunk 15]
```diff
@@ -85,55 +170,104 @@ if !([ -f /etc/ros/rosdep/sources.list.d/20-default.list ]); then
...
+catkin_make -C $CATKIN_DIR -B
```
Comment (sentree) [inline install.sh:208]: This build should only be done if the workspace has just been created, otherwise it will rebuild the user's existing workspace from scratch.

File: install.sh (hunk 12 above)
Comment (sentree) [reply install.sh:157]: Installing only ros-indigo-desktop-full on Semaphore saved a lot of time during builds, so I will have to add that back in.

File: install.sh [hunk 16]
```diff
@@ -85,55 +170,104 @@ if !([ -f /etc/ros/rosdep/sources.list.d/20-default.list ]); then
...
+git clone https://github.com/txros/txros.git
 
-# Attempt to build the sub from scratch on client machines
-if !(env | grep --quiet "SEMAPHORE=true"); then
-    instlog "Building the sub's software stack with catkin_make"
-    catkin_make -C "$CATKIN_DIR" -j8
-fi
+mv $CATKIN_DIR/src/Sub8/gnc/sub8_trajectory_generator $CATKIN_DIR/src/Sub8/gnc/.sub8_trajectory_generator
+
+catkin_make -C $CATKIN_DIR -B
```
Comment (sentree) [inline install.sh:273]: This line would result in building the repository twice on semaphore without a check.

File: install.sh [hunk 17]
```diff
@@ -85,55 +170,104 @@ if !([ -f /etc/ros/rosdep/sources.list.d/20-default.list ]); then
...
+sudo apt-get install -qq sshfs
+sudo apt-get install -qq git-lfs gitk
+git lfs install --skip-smudge
+sudo apt-get install -qq tmux
+
+# Libraries needed by txros
+sudo apt-get install -qq python-twisted socat
+
+# Package management
+sudo pip install -q -U setuptools
```
Comment (sentree) [inline install.sh:248]: You never install python-pip, which is not installed on Ubuntu server by default, so any pip installation would fail.

File: install.sh [hunk 18]
```diff
@@ -85,55 +170,104 @@ if !([ -f /etc/ros/rosdep/sources.list.d/20-default.list ]); then
...
+    ./configure
+    make
+    make check
+    sudo make install
+    cd ..
+    sudo rm -r protobuf-2.6.1
+    sudo rm -r protobuf-2.6.1.tar.gz
+else
+    instlog "Detected correct version of protoc"
+fi
```
Comment (sentree) [inline install.sh:193]: Are we sure there isn't a PPA for this?

File: install.sh [hunk 19]
```diff
@@ -85,55 +170,104 @@ if !([ -f /etc/ros/rosdep/sources.list.d/20-default.list ]); then
...
+# Machine Learning
+sudo pip install -q -U scikit-learn > /dev/null 2>&1
+
+# Visualization
+sudo pip install -q -U mayavi > /dev/null 2>&1
+
+instlog "Cloning common GIT repositories"
 cd $CATKIN_DIR/src
-$CATKIN_DIR/src/Sub8/scripts/get_dependencies.sh
+git clone https://github.com/RustyBamboo/camera1394.git
```
Comment (sentree) [inline install.sh:268]: Why is this necessary as opposed to installing ros-kinetic-camera1394?

File: install.sh [hunk 20]
```diff
@@ -85,55 +170,104 @@ if !([ -f /etc/ros/rosdep/sources.list.d/20-default.list ]); then
...
 
-# Update the hosts file if necessary
-if (env | grep --quiet "SUB=true"); then
-    $CATKIN_DIR/src/Sub8/scripts/update-hosts.bash
+# Adds Sub8 aliases to $BASHRC_FILE unless already exists
+if !(cat $BASHRC_FILE | grep --quiet "source $ALIASES_FILE"); then
+    echo "" >> $BASHRC_FILE
+    echo "# Adds Sub8 aliases to shell environment" >> $BASHRC_FILE
+    echo "source $ALIASES_FILE" >> $BASHRC_FILE
 fi
```
Comment (sentree) [inline install.sh:228]: I didn't see you move the Sub8 repository into the catkin workspace for Semaphore. If this step is not done, the master branch of Sub8 is what will be built, not the branch that was last updated.

**Review by sentree — CHANGES_REQUESTED** [review]
Since we are moving toward a software-common repository, we are trying to build a unified MIL install script. I will be building and testing one on that repository over the next week and I will commit the changes from this PR that work well with you as the author. There are still a few questions I have before I add the changes, though.

File: install.sh (hunk 7 above)
Comment (RustyBamboo) [reply install.sh:14]: Standard is probably better

File: install.sh (hunk 10 above)
Comment (RustyBamboo) [reply install.sh:54]: I think it's a good thing to have

File: install.sh (hunk 8 above)
Comment (RustyBamboo) [reply install.sh:48]: I saw no real need to add more complexity. "git clone bleh" done. Although, it may be worth to add it back for the recursive option.

File: install.sh (hunk 9 above)
Comment (RustyBamboo) [reply install.sh:59]: Woops.

File: install.sh (hunk 18 above)
Comment (RustyBamboo) [reply install.sh:193]: The PPA is for the newer version (3.0.1 - i think.) The problem is that all the gazebo protoc messages don't specify that they're using 3.0.1, and so they default to 2.6.1.

File: install.sh (hunk 14 above)
Comment (RustyBamboo) [reply install.sh:207]: The driver_base in the drive_common is needed for the Camera1394 package.

Yeah, that seems to be the case...

File: install.sh (hunk 15 above)
Comment (RustyBamboo) [reply install.sh:208]: Yeah... the problem was that the Camera1394 depends on the driver_base... But during compilation Camera1394 would fail to build because it was looking for driver_base dependences which didn't exist yet. No clue how to fix this... My solution was to build driver_common first. Maybe build driver_common into the /ros/kinetic

File: install.sh (hunk 19 above)
Comment (RustyBamboo) [reply install.sh:268]: http://wiki.ros.org/camera1394... There doesn't seem to be support for kinetic...

File: install.sh (hunk 20 above)
Comment (RustyBamboo) [reply install.sh:228]: Ah... Smart.

File: install.sh (hunk 16 above)
Comment (RustyBamboo) [reply install.sh:273]: Semaphore is gud.

Not sure what the solution is for the camera1394 depending on driver_common without building twice...

File: install.sh (hunk 7 above)
Comment (RustyBamboo) [reply install.sh:14]: Although convenience is also good, which is probably better than following standards

File: install.sh (hunk 11 above)
Comment (DSsoto) [reply install.sh:65]: Make sure you keep any specific to Sub8 in that file.

Comment (sentree) [reply install.sh:65]: That's the plan, a common aliases file and then one for each project installed.

**Comment by DSsoto** [comment]
This was a great starting point but we have decided to move all of our install scripts to the installer repo.
Closing.
